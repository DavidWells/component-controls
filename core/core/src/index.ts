export * from './utils';
export { randomizeData } from './randomizeData';
export * from './smartControls';
export * from './source';
export * from './faker';
