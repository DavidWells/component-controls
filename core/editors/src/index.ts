export * from './editors/prop-factory';
export * from './editors/types';
export * from './forms/ControlsTable/ControlEditorsTable';
export { FlexContainer } from './components/FlexContainer/FlexContainer';
