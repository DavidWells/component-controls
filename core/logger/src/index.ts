import chalk from 'chalk';

export type LogLevel = 'all' | 'errors' | 'none';

export interface LogOptions {
  logLevel?: LogLevel;
  colors: [number, number, number];
}

const defaultOptions: LogOptions = {
  logLevel: 'all',
  colors: [244, 147, 66],
};

let logOptions: LogOptions = { ...defaultOptions };

const output = (heading: string, text: string) => {
  console.log(chalk.bgRgb(...logOptions.colors)(`@${heading}`), text);
};

export const log = (heading: string, text: string): void => {
  if (logOptions.logLevel === 'all') {
    output(heading, text);
  }
};
export const error = (heading: string, text: string): void => {
  if (logOptions.logLevel !== 'none') {
    output(heading, text);
  }
};

export const setLogOptions = (options: Partial<LogOptions> = {}): void => {
  logOptions = { ...logOptions, ...options };
};
