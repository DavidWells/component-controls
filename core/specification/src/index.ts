export * from './types';

export type SetControlValueFn = (
  storyId: string,
  propertyName: string | undefined,
  value: any,
) => void;
export type ClickControlFn = (storyId: string, propertyName: string) => void;
