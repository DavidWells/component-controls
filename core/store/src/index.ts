export * from './Store/Store';
export * from './Store/HMRStore';
export * from './types';
export * from './hooks';
export * from './serialization/load-store';
export * from './create-pages';
