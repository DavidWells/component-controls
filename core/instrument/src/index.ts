import * as parser from '@babel/parser';
import * as fs from 'fs';
import mdx from '@mdx-js/mdx';
import matter from 'gray-matter';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import deepMerge from 'deepmerge';
import { Story, StoriesDoc, getASTSource } from '@component-controls/core';

import { extractCSFStories } from './babel/csf-stories';
import { extractMDXStories } from './babel/mdx-stories';
import { removeMDXAttributes } from './babel/remove-mdx-attributes';
import { extractStoreComponent } from './babel/extract-component';
import { packageInfo } from './misc/package-info';
import { extractStoryExports } from './misc/mdx-exports';
import { prettify } from './misc/prettify';
import {
  LoadingDocStore,
  InstrumentOptions,
  ParserOptions,
  defaultParserOptions,
  ResolveOptions,
  defaultResolveOptions,
  ComponentOptions,
  defaultComponentOptions,
  StoriesOptions,
  defaultStoriesOptions,
  MDXOptions,
  defaultMDXOptions,
  ParseStorieReturnType,
} from './types';

export * from './types';

type TraverseFn = (
  ast: File,
  options: Required<InstrumentOptions>,
  more: { source: string; filePath: string },
) => ParseStorieReturnType;

/**
 * Instrument a source file, with options
 * @param code The source code
 * @param traverseFn A traverse function. can be different for MDX and CSF story files
 * @param originalSource If the source was modified (ie mdx compiler)
 * @param filePath file name with fiull path
 * @param options Instrument options
 */
const parseSource = async (
  code: string,
  traverseFn: TraverseFn,
  originalSource: string,
  filePath: string,
  options: Required<InstrumentOptions>,
): Promise<ParseStorieReturnType> => {
  const source = await prettify(code, filePath, options.prettier);
  const ast = parser.parse(source, options.parser);

  const store = traverseFn(ast, options, { source, filePath });
  if (store.doc) {
    const doc = store.doc;
    if (options.stories.storeSourceFile) {
      doc.source = originalSource;
    }
  }
  await extractStoreComponent(store, filePath, source, options, ast);
  const doc: StoriesDoc | undefined = store.doc;
  if (doc && store.stories) {
    let includedStories = Object.keys(store.stories);
    const { includeStories, excludeStories } = doc;
    if (includeStories) {
      if (Array.isArray(includeStories)) {
        includedStories = includedStories.filter(
          name => includeStories.indexOf(name) > -1,
        );
      } else {
        const match = new RegExp(includeStories);
        includedStories = includedStories.filter(name => {
          return name.match(match);
        });
      }
    }
    if (excludeStories) {
      if (Array.isArray(excludeStories)) {
        includedStories = includedStories.filter(
          name => excludeStories.indexOf(name) === -1,
        );
      } else {
        const match = new RegExp(excludeStories);
        includedStories = includedStories.filter(name => !name.match(match));
      }
    }
    if (includeStories || excludeStories) {
      for (var key in store.stories) {
        if (includedStories.indexOf(key) === -1) {
          delete store.stories[key];
        }
      }
    }
    const storyPackage = await packageInfo(filePath, options.stories.package);
    if (storyPackage) {
      store.packages[storyPackage.fileHash] = storyPackage;
      doc.package = storyPackage.fileHash;
    }
    const stats = fs.statSync(filePath);
    doc.dateModified = doc.dateModified || stats.mtime;
    doc.date = doc.date || stats.birthtime;
  }
  for (const key of Object.keys(store.stories)) {
    const story: Story = store.stories[key];
    if (!story.source) {
      story.source = getASTSource(source, story.loc);
    }
  }
  return store;
};

export { LoadingDocStore };
export type ParseStoriesReturnType = { transformed: string } & LoadingDocStore;
/**
 * Parse and instrument a javascript, typescript or MDX file of stories
 * @param source Source of the file to be instrumented
 * @param filePath Resolved file path name.
 * @param options Instrumenting options
 */

export const parseStories = async (
  source: string,
  filePath: string,
  options?: InstrumentOptions,
): Promise<ParseStoriesReturnType> => {
  const {
    parser: parserOptions = {},
    prettier: prettierOptions = {},
    resolver: resolveOptions = {},
    components: componentOptions = {},
    stories: storiesOptions = {},
    mdx: mdxOptions = {},
    propsLoaders,
  } = options || {};

  const mergedOptions: Required<InstrumentOptions> = {
    parser: deepMerge<ParserOptions>(defaultParserOptions, parserOptions),
    resolver: deepMerge<ResolveOptions>(defaultResolveOptions, resolveOptions),
    prettier: prettierOptions,
    components: deepMerge<ComponentOptions>(
      defaultComponentOptions,
      componentOptions,
    ),
    stories: deepMerge<StoriesOptions>(defaultStoriesOptions, storiesOptions),
    mdx: deepMerge<MDXOptions>(defaultMDXOptions, mdxOptions),
    propsLoaders: propsLoaders || [],
  };
  let code: string;
  const {
    test,
    renderer,
    transformMDX,
    ...otherMDXOptions
  } = mergedOptions.mdx;
  if (test && filePath.match(test)) {
    const { data, content } = matter(source);
    const mdxParsed = await mdx(content, {
      filepath: filePath,
      ...otherMDXOptions,
    });
    const ast = parser.parse(mdxParsed, mergedOptions.parser) as any;
    traverse(ast, removeMDXAttributes());
    ({ code } = generate(ast, {
      retainFunctionParens: true,
      retainLines: true,
    }));
    const store = await parseSource(
      code,
      extractMDXStories(data),
      content,
      filePath,
      mergedOptions,
    );
    const { stories, doc, components, exports, packages } = store;
    const exportsSource = extractStoryExports(exports);
    let transformed = `
    
    ${content}
`;
    if (transformMDX) {
      transformed = `${renderer}\n${code}\n${exportsSource}`;
    }
    return {
      transformed,
      stories,
      doc,
      components,
      packages,
    };
  } else {
    const store = await parseSource(
      source,
      extractCSFStories,
      source,
      filePath,
      mergedOptions,
    );
    return {
      transformed: source,
      ...store,
    };
  }
};
