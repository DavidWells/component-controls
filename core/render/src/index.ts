import { FrameworkRenderFn } from '@component-controls/core';
import { render as renderReact } from './react';

//noop
export const renderers: Record<string, FrameworkRenderFn> = {
  react: renderReact,
};
