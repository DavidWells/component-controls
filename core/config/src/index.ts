import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
