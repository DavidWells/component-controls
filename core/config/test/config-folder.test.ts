import { getConfiguration } from '../src/index';

describe('config-folder', () => {
  it('config file short option', () => {
    expect(
      getConfiguration(__dirname, ['file', 'name', '-c', 'fixtures']),
    ).toMatchSnapshot();
  });
  it('config file long option', () => {
    expect(
      getConfiguration(__dirname, ['file', 'name', '--config', 'fixtures']),
    ).toMatchSnapshot();
  });
});
