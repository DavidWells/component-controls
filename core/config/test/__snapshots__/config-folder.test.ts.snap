// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config-folder config file long option 1`] = `
Object {
  "config": Object {
    "addons": Array [
      Object {
        "name": "@component-controls/storybook",
        "options": Object {
          "webpack": Array [
            "instrument",
            "react-docgen-typescript",
          ],
        },
      },
    ],
    "stories": Array [
      "../../../../ui/editors/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../ui/components/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../ui/blocks/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../examples/stories/src/**/*.stories.(js|jsx|tsx|mdx)",
    ],
  },
  "configPath": "/Users/atanasster/component-controls/core/config/test/fixtures",
  "optionsFilePath": undefined,
}
`;

exports[`config-folder config file short option 1`] = `
Object {
  "config": Object {
    "addons": Array [
      Object {
        "name": "@component-controls/storybook",
        "options": Object {
          "webpack": Array [
            "instrument",
            "react-docgen-typescript",
          ],
        },
      },
    ],
    "stories": Array [
      "../../../../ui/editors/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../ui/components/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../ui/blocks/src/**/*.stories.(js|jsx|tsx|mdx)",
      "../../../../examples/stories/src/**/*.stories.(js|jsx|tsx|mdx)",
    ],
  },
  "configPath": "/Users/atanasster/component-controls/core/config/test/fixtures",
  "optionsFilePath": undefined,
}
`;
